Ensure empty string path is properly handled, add --all-paths flag

riyazdf opened the pull request:
This improves notary CLI to properly display `""` in the console (before it was just blank) and also describes it as `<all paths>`.  This also adds a `--all-paths` flag that will add `""` when used with`delegation add`, and clear all paths on `delegation remove`.  Removing the individual `""` path will only remove that path.

Closes #505.

Signed-off-by: Riyaz Faizullabhoy <riyaz.faizullabhoy@docker.com>

cyli commented:
This definitely makes everything a lot easier!  :D  Unfortunately I think it may need to take into account whether we are modifying a role or not?  I tried the following:

1. `notary init repo1`
2. `notary delegation add repo1 targets/releases <path-to-pem>`
3. `notary publish repo1`

Awesome, after updating the repo, the delegation has the empty path.

But if I do:

1. `notary init repo2`
2. `notary delegation add repo2 targets/releases <path-to-pem> --paths="/"`
3. `notary publish repo2`
4. `notary delegation add repo2 targets/releases <path-to-pem2>`
5. `notary publish repo2`

I get a delegation with both keys, but paths "/" and "", and "" is more permissive so it might take precedence.  I'm not sure that's desirable?  Maybe you had it right in the first place (nil by default) and we just have to better document how to add the empty string?  Not sure.  :|  @endophage?

riyazdf commented:
@cyli You're definitely right that it makes modifying a pre-existing delegation less intuitive.  I can't think of a good way to get around this unless we go with your suggestion of making `nil` the default, and documenting how to input empty string (which has to be escaped as `--paths=\"\"`).  I'm definitely open to ideas here, I'm leaning toward the `nil` default and stronger documentation -- maybe this can be made clearer in the output after running `notary delegation add`?

A second observation from your example: if a delegation already has a path of `/a`, and a user tries to add path `/a/b`, should we even bother showing `/a/b` if it is a successful publish?  We could probably try to do some smarter merging when adding to the delegation (right now we just merge string slices as is).

cyli commented:
@riyazdf I remember @endophage mentioning something like that before, where we were debating whether we wanted to check the prefixes of the paths when merging, and maybe the outcome was that it should probably be resolved by the admin?.  It would probably be a nice UI feature for the CLI to log a warning or something if it detects that, though.

endophage commented:
Yeah, inefficient/overlapping paths should be resolved by admin. Safer than us trying to be clever.

On the empty path problem. Are we able to move the addition of the empty path to the time at which we apply the changelist? At that point I believe we know whether the role already exists or not, so we can detect if we're creating a role, and if no specific paths have been supplied.

riyazdf commented:
@endophage We can detect if the role already exists when we apply the changelist, but we won't know if the user passed in `--paths=\"\"` or if the flag defaulted to `""`.

We could get around this by defaulting the `--paths` flag to `nil` and we default to a `""` path when applying changelists if we get a `nil` path on a brand new delegation?  What do you (and @cyli) think?  We'll need to make it very clear that we add `""` by default to new delegations without specified paths.

diogomonica commented:
`--all-paths`

riyazdf commented:
I added in the `--all-paths` flag, which required an additional key to the embedded delegation object in the changelist.

I also changed `""` to be printed as `"" <all paths>` on the console to make things clearer when adding all paths, but I can change this back if we'd prefer not to.

endophage commented:
Great idea on the console printing. Users accommodate with a lot of non-intuitive stuff when it's well messaged.

diogomonica commented on client/changelist/change.go:
```diff
@@ -88,6 +88,7 @@ type TufDelegation struct {
 	RemoveKeys             []string     `json:"remove_keys,omitempty"`
 	AddPaths               []string     `json:"add_paths,omitempty"`
 	RemovePaths            []string     `json:"remove_paths,omitempty"`
+	ClearPaths             bool         `json:"clear_paths,omitempty"`
```
`ClearAllPaths`

riyazdf replied:
:+1:

diogomonica commented on client/client.go:
```diff
@@ -349,7 +349,7 @@ func (r *NotaryRepository) AddDelegation(name string, threshold int,
 // This does not validate that the delegation exists, since one might exist
 // after applying all changes.
-func (r *NotaryRepository) RemoveDelegation(name string, keyIDs, paths []string, removeAll bool) error {
+func (r *NotaryRepository) RemoveDelegation(name string, keyIDs, paths []string, clearPaths, removeAll bool) error {
```
`clearAllPaths`

diogomonica commented on client/client.go:
```diff
@@ -378,6 +378,7 @@ func (r *NotaryRepository) RemoveDelegation(name string, keyIDs, paths []string,
 		tdJSON, err := json.Marshal(&changelist.TufDelegation{
 			RemoveKeys:  keyIDs,
 			RemovePaths: paths,
+			ClearPaths:  clearPaths,
```
`ClearAllPaths`

diogomonica commented on client/helpers.go:
```diff
@@ -121,7 +121,12 @@ func changeTargetsDelegation(repo *tuf.Repo, c changelist.Change) error {
-		r.RemovePaths(td.RemovePaths)
+
+		removePaths := td.RemovePaths
+		if td.ClearPaths {
```
Not sure if I'm convinced of the need of adding a new `bool`. What if we iterate through `removePaths`, and if we find a `""`, we simply remove all of them?

riyazdf replied:
We could smartly remove paths (and it would be simpler code), but we would have to make it clear that there is smarter path logic around removing `""` to users, since we don't do any smart path logic anywhere else.  Users can currently add `""` and `path` as paths to the same delegation and both will display on the CLI, even though it doesn't really make sense in notary.

cc @endophage since we were talking earlier about how to manage paths via CLI

endophage replied:
@diogomonica 's suggesting is too dependent on order of operations. If a user wants to take an existing role that has the `""` path permissions, and restrict it to `"foo/"`, this implicit behaviour would require them to remove `""` before adding `"foo/"`.

diogomonica commented on cmd/notary/delegations.go:
```diff
@@ -160,7 +165,7 @@ func (d *delegationCommander) delegationRemove(cmd *cobra.Command, args []string
 // delegationAdd creates a new delegation by adding a public key from a certificate to a specific role in a GUN
 func (d *delegationCommander) delegationAdd(cmd *cobra.Command, args []string) error {
-	if len(args) < 2 || len(args) < 3 && paths == nil {
+	if len(args) < 2 || len(args) < 3 && addPaths == nil && !allPaths {
```
I think it needs a comment on intended behavior

riyazdf replied:
good idea, will include

diogomonica commented on cmd/notary/delegations.go:
```diff
@@ -188,6 +193,11 @@ func (d *delegationCommander) delegationAdd(cmd *cobra.Command, args []string) e
 	}
 
+	// If the user passes --all-paths, give the "" path
+	if allPaths {
```
What if we special case this to remove all the current paths when `""` gets added?

riyazdf replied:
I think this is similar to this: https://github.com/docker/notary/pull/514/files#r51305926

I think we should decide whether or not we should have any smart logic around adding/removing paths.  It would require changes to `AddPaths` and `RemovePaths` in `roles.go` -- right now they just merge/subtract string slices without any logic around the values of the included paths.

diogomonica replied:
It might be ok to only do "smart logic" around the "all paths" exception. If you add `""` and other stuff, ignore the other stuff (or remove it). If you delete `""` also make sure everything else is empty.

We don't have to do more smart logic than that. Not sure what @endophage things, but I would rather do that VS having another bool on the api.

riyazdf replied:
I agree that this is a nice compromise, `""` is a special prefix in this case and I agree it's better than having to add more object keys and booleans :)  @endophage and @cyli what do you think?

The only downside I can see here is if someone mistakenly adds `""` when they didn't mean to, and loses track of the previous paths, but this is a small headache and easily recoverable.

riyazdf commented:
I just implemented the "smart logic" that @diogomonica outlined around the `""` path (`""` clears other paths that are passed in when adding, removing `""` removes all paths) because I think it's the best way to use notary here, and doesn't further complicate the API.

I also think that we can think about tweaking the logic around removing the `""` path in particular once we improve the API into finer-grained calls https://github.com/docker/notary/issues/525.  Once we have a simpler way of clearing paths vs. removing just the `""` path, the only way to accomplish this is by adding more flags to `RemoveDelegation`.

diogomonica commented:
Rebase needed.

riyazdf commented:
@diogomonica rebased!

diogomonica commented:
@endophage final review?

endophage commented on cmd/notary/delegations.go:
```diff
@@ -38,8 +38,8 @@ var cmdDelegationAddTemplate = usageTemplate{
 
-var paths []string
-var removeAll, removeYes bool
+var addPaths, removePaths []string
```
What's the necessity for these 2 separate vars? Also:

```
var (
    paths []string
    allPaths, removeAll, removeYes bool
)
```

riyazdf replied:
I'll make them one var -- earlier they had different default values and so we needed two separate variables.  I'll also reformat this as you suggested!

endophage commented on cmd/notary/delegations.go:
```diff
@@ -38,8 +38,8 @@ var cmdDelegationAddTemplate = usageTemplate{
-var paths []string
-var removeAll, removeYes bool
+var addPaths, removePaths []string
+var allPaths, removeAll, removeYes bool
```
We probably want to use the `-y` flag in other places (or will do), so I'd just call it `forceYes` or something more generic than `removeYes`

riyazdf replied:
Makes sense, will do

endophage commented:
Sorry to be the naysayer but I still completely disagree with the remove command using `""` as an alternative  to infer "remove all".

Also wondering whether the flags should be `allowAll` and `denyAll` rather than `allPaths`. It's clearer about the behaviour whereas `allPaths` is intuitive to us only because we understand the data structures involved.

diogomonica commented:
@endophage your alternative being?

endophage commented:
The same thing I was suggesting last week. We already have an explicit flag, there's no need for this implicit behaviour, just treat `""` like it was any other path. If it exists in the role's `Path` list it gets removed, if it doesn't, it's a no-op

riyazdf commented:
@endophage I think that's a fine idea but I'm not sure if I see the benefit of allowing for this behavior against adding a flag to the `RemoveDelegation` API, since `notary delegation remove` is an offline operation.  I think making the API simpler and more flexible (see https://github.com/docker/notary/issues/525) will improve this and maybe we could allow for this behavior then.

In any case, I don't think this special way of handling `""` on remove takes away any functionality from the user -- instead of being able to add paths and then remove `""` individually, we're just making the user remove `""` first before adding paths he wants to restrict the delegation to.

diogomonica commented:
Agree with @riyazdf I think we should refactor.

diogomonica commented:
@riyazdf this needs rebase after #528 got merged.

riyazdf commented:
@diogomonica @endophage refactored and rebased.

Now `notary delegation remove --all-paths` will clear all paths, and `notary delegation remove --paths=\"\"` will just remove the empty path.  I've added these to the integration test.

diogomonica commented on cmd/notary/integration_test.go:
```diff
@@ -365,6 +385,86 @@ func TestClientDelegationsInteraction(t *testing.T) {
+	assert.NoError(t, err)
+
+	// Add more individual paths so we can test a delegation remove --all-paths
+	output, err = runCommand(t, tempDir, "delegation", "add", "gun", "targets/delegation", "--paths", "banana,apple,orange,kiwi")
```
maybe some more complex paths? `banana/split` comes to mind ;)

diogomonica commented on cmd/notary/prettyprint.go:
```diff
@@ -198,6 +198,19 @@ func prettyPrintRoles(rs []*data.Role, writer io.Writer, roleType string) {
+	for _, path := range paths {
+		// manually escape "" and designate that it is all paths with an extra print <all paths>
+		if path == "" {
+			path = "\"\" <all paths>"
```
Why include `""`?

riyazdf replied:
I think we should still display the actual `""` path so that the `notary delegation remove --paths=\"\"` has meaning, what do you think?

diogomonica replied:
Sounds good.

cyli replied:
Totally non-blocking, and this may already be done elsewhere, but I probably missed it - would sorting the paths make it easier for a user to see if they have overlapping path permissions?  So the all paths would always be in front, and then paths that overlap each other would be right next to each other?

endophage replied:
I like @cyli's suggestion. Sort in lexical order, maybe we can even add some output syntax to indicate shadowing?

riyazdf replied:
Great suggestion!

diogomonica commented:
LGTM

diogomonica commented:
@riyazdf oops, sorry about this. I realized that this was going to need a rebase after I merged @cyli PR.

riyazdf commented:
@diogomonica no problem, just rebased.

diogomonica commented on cmd/notary/integration_test.go:
```diff
@@ -362,6 +382,86 @@ func TestClientDelegationsInteraction(t *testing.T) {
+	assert.NotContains(t, output, "123456")
+	assert.NotContains(t, output, "banana/split")
+	assert.NotContains(t, output, "apple/crumble/pie")
+	assert.NotContains(t, output, "orange.peel")
```
love it

cyli commented on cmd/notary/delegations.go:
```diff
@@ -171,9 +179,17 @@ func (d *delegationCommander) delegationRemove(cmd *cobra.Command, args []string
+			removingItems = removingItems + fmt.Sprintf("with keys %s, ", keyIDs)
+		}
+		if d.allPaths {
+			removingItems = removingItems + "with all paths,"
```
Non-blocking nitpick - should this string terminate with a space like the other ones?  Alternately should none of them have a trailing space and the format string should just include the space?

riyazdf replied:
Good point!  I'll fix this.

cyli commented on cmd/notary/delegations.go:
```diff
@@ -239,9 +268,16 @@ func (d *delegationCommander) delegationAdd(cmd *cobra.Command, args []string) e
 	cmd.Printf(
-		"Addition of delegation role %s with keys %s and paths %s, to repository \"%s\" staged for next publish.\n",
-		role, pubKeyIDs, d.paths, gun)
+		"Addition of delegation role %s %s to repository \"%s\" staged for next publish.\n",
```
Non-blocking nitpick Should this be `%sto repository` like above in the `delegationRemove` command, since all the additions have trailing spaces?

riyazdf replied:
yes!  Will update

cyli commented:
Could we also two test cases where both `--path` and `--all-paths` are given simultaneously, and that if so, the `--paths` are ignored?

cyli commented:
LGTM otherwise!  :D

riyazdf commented:
@cyli added your suggested test cases to the integration test and also added sorting to path pretty printing!

cyli commented:
@riyazdf :+1: Thanks!! :D

diogomonica commented:
Great team-work. This is looking good.
